aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -95,6 +110,70 @@ struct FusedOptimizerTensorListMetadata {
   int start_tensor_this_launch;
 };
 
+bool can_use_static_tensor_list_meta(
+    std::vector<std::vector<at::Tensor>>& tensor_lists,
+    int depth) {
+  const int64_t n_tensors = tensor_lists[0].size();
+  if (n_tensors > depth_to_max_tensors[depth - 1]) {
+    return false;
+  }
+  int64_t num_blocks = 0;
+  for (const auto t : c10::irange(n_tensors)) {
+    const auto numel = tensor_lists[0][t].numel();
+    const auto chunks = numel / kChunkSize + (numel % kChunkSize != 0);
+    num_blocks += chunks;
+    if (num_blocks > depth_to_max_blocks[depth - 1]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Helper for transfering multiple std::vector<T> onto device with a single
+// page-locked cudaMemcpyAsync.
+struct VecPacker {
+  template <typename T>
+  // Add a vector to be copied to device
+  // NOTE: VecPacker doesn't make copies of the added vectors. They have to be
+  // kept alive by the caller until .pack() is called.
+  void add(const std::vector<T>& vec) {
+    ptrs.push_back(vec.data());
+    const auto vec_bytes = sizeof(T) * vec.size();
+    const auto vec_bytes_aligned = (vec_bytes + 16 - 1) / 16 * 16;
+    sizes.push_back(vec_bytes);
+    offsets.push_back(packed_numel);
+    packed_numel += vec_bytes_aligned;
+  }
+
+  // Copy all previously added vectors onto device and return their device
+  // pointers in the order they are added. We leverage the stream awareness of
+  // CUDACachingAllocator to manage the lifetime of the device arguments - the
+  // device memory is guaranteed to be alive as long as VecPacker is destoryed
+  // after the kernel that consumes it.
+  std::vector<void*> pack(const at::Device& device) {
+    packed = at::empty(
+        {packed_numel},
+        at::TensorOptions().dtype(at::kByte).pinned_memory(true));
+    for (const auto i : c10::irange(ptrs.size())) {
+      memcpy(packed.data_ptr<uint8_t>() + offsets[i], ptrs[i], sizes[i]);
+    }
+    packed = packed.to(device, /*non_blocking=*/true);
+
+    std::vector<void*> dev_ptrs;
+    dev_ptrs.reserve(ptrs.size());
+    for (const auto offset : offsets) {
+      dev_ptrs.push_back(packed.data_ptr<uint8_t>() + offset);
+    }
+    return dev_ptrs;
+  }
+
+  std::vector<const void*> ptrs;
+  std::vector<size_t> sizes;
+  std::vector<size_t> offsets;
+  int64_t packed_numel = 0;
+  at::Tensor packed;
```

Comment:
Style q: is it typical to put the struct attributes beneath the function defs? I think I'm more used to these being above cuz I find it easier to read top-down but I don't actually know the style recommendation.

Reply:
I'm not familiar with any. But now that you mention it, it does look better to put them on top :)